[TEST-FAILURE] InvocationNestedRemoteTest.whenPartition_callsIncorrectPartition

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.fail(Assert.java:95)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationNestedRemoteTest.whenPartition_callsIncorrectPartition(InvocationNestedRemoteTest.java:105)
```
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OracleJDK1.7/com.hazelcast$hazelcast/166/testReport/junit/com.hazelcast.spi.impl.operationservice.impl/InvocationNestedRemoteTest/whenPartition_callsIncorrectPartition/

--- a/checkstyle/suppressions.xml
+++ b/checkstyle/suppressions.xml
@@ -277,6 +277,7 @@
     <suppress checks="CyclomaticComplexity" files="com/hazelcast/spi/impl/operationservice/impl/InvocationFuture"/>
     <suppress checks="ClassDataAbstractionCoupling|MethodCount"
               files="com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl"/>
+    <suppress checks="ReturnCount" files="com/hazelcast/spi/impl/operationexecutor/classic/ClassicOperationExecutor"/>
 
     <!-- Transaction -->
     <suppress checks="JavadocMethod" files="com/hazelcast/transaction/"/>

--- a/hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/classic/ClassicOperationExecutor.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/classic/ClassicOperationExecutor.java
@@ -186,13 +186,13 @@ public final class ClassicOperationExecutor implements OperationExecutor {
         return threads;
     }
 
-    @SuppressFBWarnings({"EI_EXPOSE_REP" })
+    @SuppressFBWarnings({ "EI_EXPOSE_REP" })
     @Override
     public OperationRunner[] getPartitionOperationRunners() {
         return partitionOperationRunners;
     }
 
-    @SuppressFBWarnings({"EI_EXPOSE_REP" })
+    @SuppressFBWarnings({ "EI_EXPOSE_REP" })
     @Override
     public OperationRunner[] getGenericOperationRunners() {
         return genericOperationRunners;
@@ -248,18 +248,24 @@ public final class ClassicOperationExecutor implements OperationExecutor {
             return true;
         }
 
+        // allowed to invoke non partition specific task
         if (op.getPartitionId() < 0) {
             return true;
         }
 
-        // we are allowed to invoke from non PartitionOperationThreads (including GenericOperationThread)
+        // allowed to invoke from non PartitionOperationThreads (including GenericOperationThread)
         if (!(currentThread instanceof PartitionOperationThread)) {
             return true;
         }
 
-        // we are only allowed to invoke from a PartitionOperationThread if the operation belongs to that
-        // PartitionOperationThread.
         PartitionOperationThread partitionThread = (PartitionOperationThread) currentThread;
+        OperationRunner runner = partitionThread.getCurrentOperationRunner();
+        if (runner != null) {
+            // non null runner means it's a nested call
+            // in this case partitionId of both inner and outer operations have to match
+            return runner.getPartitionId() == op.getPartitionId();
+        }
+
         return toPartitionThreadIndex(op.getPartitionId()) == partitionThread.threadId;
     }
 

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/InvocationFutureTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/InvocationFutureTest.java
@@ -36,14 +36,11 @@ public class InvocationFutureTest extends HazelcastTestSupport {
         operationService = getOperationService(local);
     }
 
-    // ================= isDone ==========================================
-
     @Test
     public void isDone_whenNullResponse() throws ExecutionException, InterruptedException {
         DummyOperation op = new DummyOperation(null);
 
         InternalCompletableFuture future = operationService.invokeOnTarget(null, op, getAddress(local));
-        // first we wait for the future to complete.
         future.get();
 
         assertTrue(future.isDone());
@@ -51,76 +48,40 @@ public class InvocationFutureTest extends HazelcastTestSupport {
 
     @Test
     public void isDone_whenWaitResponse() {
-        DummyOperation op = new DummyOperation() {
-            {
-                // we need to set the call-id to prevent running the operation on the calling-thread.
-                setPartitionId(1);
-            }
-
-            @Override
-            public void run() throws Exception {
-                Thread.sleep(5000);
-            }
-        };
+        DummyOperation op = new GetLostPartitionOperation();
 
         InvocationFuture future = (InvocationFuture) operationService.invokeOnTarget(null, op, getAddress(local));
         future.set(InternalResponse.WAIT_RESPONSE);
+
         assertFalse(future.isDone());
     }
 
     @Test
     public void isDone_whenInterruptedResponse() {
-        DummyOperation op = new DummyOperation() {
-            {
-                // we need to set the call-id to prevent running the operation on the calling-thread.
-                setPartitionId(1);
-            }
-
-            @Override
-            public void run() throws Exception {
-                Thread.sleep(5000);
-            }
-        };
+        DummyOperation op = new GetLostPartitionOperation();
 
         InvocationFuture future = (InvocationFuture) operationService.invokeOnTarget(null, op, getAddress(local));
         future.set(InternalResponse.INTERRUPTED_RESPONSE);
+
         assertTrue(future.isDone());
     }
 
     @Test
     public void isDone_whenTimeoutResponse() {
-        DummyOperation op = new DummyOperation() {
-            {
-                // we need to set the call-id to prevent running the operation on the calling-thread.
-                setPartitionId(1);
-            }
-
-            @Override
-            public void run() throws Exception {
-                Thread.sleep(5000);
-            }
-        };
+        DummyOperation op = new GetLostPartitionOperation();
 
         InvocationFuture future = (InvocationFuture) operationService.invokeOnTarget(null, op, getAddress(local));
         future.set(InternalResponse.TIMEOUT_RESPONSE);
+
         assertTrue(future.isDone());
     }
 
     @Test
     public void isDone_whenNoResponse() {
-        DummyOperation op = new DummyOperation() {
-            {
-                // we need to set the call-id to prevent running the operation on the calling-thread.
-                setPartitionId(1);
-            }
-
-            @Override
-            public void run() throws Exception {
-                Thread.sleep(5000);
-            }
-        };
+        DummyOperation op = new GetLostPartitionOperation();
 
         InternalCompletableFuture future = operationService.invokeOnTarget(null, op, getAddress(local));
+
         assertFalse(future.isDone());
     }
 
@@ -129,22 +90,19 @@ public class InvocationFutureTest extends HazelcastTestSupport {
         DummyOperation op = new DummyOperation("foobar");
 
         InternalCompletableFuture future = operationService.invokeOnTarget(null, op, getAddress(local));
+
         assertTrue(future.isDone());
     }
-
-    // ========================= andThen ==================================
 
     // There is a bug: https://github.com/hazelcast/hazelcast/issues/5001
     @Test
     public void andThen_whenNullResponse_thenCallbackExecuted() throws ExecutionException, InterruptedException {
         DummyOperation op = new DummyOperation(null);
-
+        final ExecutionCallback callback = mock(ExecutionCallback.class);
         InternalCompletableFuture future = operationService.invokeOnTarget(null, op, getAddress(local));
-        // first we wait for the future to complete.
         future.get();
 
-        // if we now register a callback, it could complete immediately since a response already has been set (NULL_RESPONSE)
-        final ExecutionCallback callback = mock(ExecutionCallback.class);
+        // Callback can be completed immediately since a response (NULL_RESPONSE) has been already set.
         future.andThen(callback);
 
         assertTrueEventually(new AssertTask() {
@@ -154,4 +112,18 @@ public class InvocationFutureTest extends HazelcastTestSupport {
             }
         });
     }
+
+    // Needed to have an invocation and this is the easiest way how to get one and do not bother with its result.
+    private static class GetLostPartitionOperation extends DummyOperation {
+        {
+            // we need to set the call-id to prevent running the operation on the calling-thread.
+            setPartitionId(1);
+        }
+
+        @Override
+        public void run() throws Exception {
+            Thread.sleep(5000);
+        }
+    }
+
 }

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/InvocationNestedLocalTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/InvocationNestedLocalTest.java
@@ -8,115 +8,120 @@ import com.hazelcast.spi.OperationService;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
 import org.junit.runner.RunWith;
 
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.Future;
 
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.fail;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
 public class InvocationNestedLocalTest extends InvocationNestedTest {
 
-    private final String response = "someresponse";
+    private static final String RESPONSE = "someresponse";
+
+    @Rule
+    public ExpectedException expected = ExpectedException.none();
 
     @Test
-    public void whenPartition_callsGeneric() {
-        HazelcastInstance hz = createHazelcastInstance();
-        OperationService operationService = getOperationService(hz);
+    public void invokeOnPartition_outerGeneric_innerGeneric_forbidden() {
+        HazelcastInstance local = createHazelcastInstance();
+        OperationService operationService = getOperationService(local);
 
-        InnerOperation innerOperation = new InnerOperation(response, -1);
-        OuterOperation outerOperation = new OuterOperation(innerOperation, 0);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, GENERIC_OPERATION);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, GENERIC_OPERATION);
 
-        InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
-
-        assertEquals(response, f.getSafely());
+        expected.expect(Exception.class);
+        operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
     }
 
     @Test
-    public void whenPartition_callsCorrectPartition() {
-        HazelcastInstance hz = createHazelcastInstance();
-        OperationService operationService = getOperationService(hz);
+    public void invokeOnPartition_outerLocal_innerGeneric() {
+        HazelcastInstance local = createHazelcastInstance();
+        OperationService operationService = getOperationService(local);
 
-        int partitionId = 0;
-        InnerOperation innerOperation = new InnerOperation(response, partitionId);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, GENERIC_OPERATION);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, getPartitionId(local));
+        InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
+
+        assertEquals(RESPONSE, f.getSafely());
+    }
+
+    @Test
+    public void invokeOnPartition_outerLocal_innerSameInstance_samePartition() {
+        HazelcastInstance local = createHazelcastInstance();
+        OperationService operationService = getOperationService(local);
+
+        int partitionId = getPartitionId(local);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, partitionId);
         OuterOperation outerOperation = new OuterOperation(innerOperation, partitionId);
-
         InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        assertEquals(response, f.getSafely());
+        assertEquals(RESPONSE, f.getSafely());
     }
 
     @Test
-    public void whenPartition_callsIncorrectPartition() {
-        HazelcastInstance hz = createHazelcastInstance();
-        OperationService operationService = getOperationService(hz);
+    public void invokeOnPartition_outerLocal_innerSameInstance_callsDifferentPartition() {
+        HazelcastInstance local = createHazelcastInstance();
+        OperationService operationService = getOperationService(local);
 
-        int outerPartitionId = 0;
-        int innerPartitionId = 1;
-        for (; innerPartitionId < hz.getPartitionService().getPartitions().size(); innerPartitionId++) {
-            if (!mappedToSameThread(operationService, outerPartitionId, innerPartitionId)) {
-                break;
-            }
-        }
-
-        InnerOperation innerOperation = new InnerOperation(response, innerPartitionId);
+        int outerPartitionId = getPartitionId(local);
+        int innerPartitionId = randomPartitionIdNotMappedToSameThreadAsGivenPartitionIdOnInstance(local, outerPartitionId);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, innerPartitionId);
         OuterOperation outerOperation = new OuterOperation(innerOperation, outerPartitionId);
-
         InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        try {
-            f.getSafely();
-            fail();
-        } catch (IllegalThreadStateException e) {
-        }
+        expected.expect(IllegalThreadStateException.class);
+        expected.expectMessage("cannot make remote call");
+        f.getSafely();
     }
 
     @Test
-    public void whenPartition_callsDifferentPartition_butMappedToSameThread() throws ExecutionException, InterruptedException {
+    public void invokeOnPartition_outerLocal_innerSameInstance_callsDifferentPartition_mappedToSameThread() throws ExecutionException, InterruptedException {
         Config config = new Config();
         config.setProperty(GroupProperty.PARTITION_COUNT, "2");
         config.setProperty(GroupProperty.PARTITION_OPERATION_THREAD_COUNT, "1");
-        HazelcastInstance hz = createHazelcastInstance(config);
-        final OperationService operationService = getOperationService(hz);
+        HazelcastInstance local = createHazelcastInstance(config);
+        final OperationService operationService = getOperationService(local);
 
-        int innerPartitionId = 0;
         int outerPartitionId = 1;
-        InnerOperation innerOperation = new InnerOperation(response, innerPartitionId);
-        OuterOperation outerOperation = new OuterOperation(innerOperation, outerPartitionId);
-
-        Future f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
-        assertEquals(response, f.get());
-    }
-
-    @Test
-    public void whenGeneric_callsGeneric() {
-        HazelcastInstance hz = createHazelcastInstance();
-        OperationService operationService = getOperationService(hz);
-
-        InnerOperation innerOperation = new InnerOperation(response, -1);
-        OuterOperation outerOperation = new OuterOperation(innerOperation, -1);
-
-        InternalCompletableFuture f = operationService.invokeOnTarget(null, outerOperation, getAddress(hz));
-
-        assertEquals(response, f.getSafely());
-    }
-
-    @Test
-    public void whenGeneric_callsPartitionSpecific() {
-        HazelcastInstance hz = createHazelcastInstance();
-        OperationService operationService = getOperationService(hz);
-
         int innerPartitionId = 0;
-        InnerOperation innerOperation = new InnerOperation(response, innerPartitionId);
-        OuterOperation outerOperation = new OuterOperation(innerOperation, -1);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, innerPartitionId);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, outerPartitionId);
+        InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        InternalCompletableFuture f = operationService.invokeOnTarget(null, outerOperation, getAddress(hz));
-
-        assertEquals(response, f.getSafely());
+        expected.expect(IllegalThreadStateException.class);
+        expected.expectMessage("cannot make remote call");
+        f.getSafely();
     }
+
+    @Test
+    public void invokeOnTarget_outerGeneric_innerGeneric() {
+        HazelcastInstance local = createHazelcastInstance();
+        OperationService operationService = getOperationService(local);
+
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, GENERIC_OPERATION);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, GENERIC_OPERATION);
+        InternalCompletableFuture f = operationService.invokeOnTarget(null, outerOperation, getAddress(local));
+
+        assertEquals(RESPONSE, f.getSafely());
+    }
+
+    @Test
+    public void invokeOnTarget_outerGeneric_innerSameInstance() {
+        HazelcastInstance local = createHazelcastInstance();
+        OperationService operationService = getOperationService(local);
+
+        int innerPartitionId = getPartitionId(local);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, innerPartitionId);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, GENERIC_OPERATION);
+        InternalCompletableFuture f = operationService.invokeOnTarget(null, outerOperation, getAddress(local));
+
+        assertEquals(RESPONSE, f.getSafely());
+    }
+
 }

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/InvocationNestedRemoteTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/InvocationNestedRemoteTest.java
@@ -6,89 +6,88 @@ import com.hazelcast.spi.OperationService;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
 import org.junit.runner.RunWith;
 
 import java.util.concurrent.ExecutionException;
 
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.fail;
+import static org.junit.Assert.assertNotEquals;
 
-/**
- * Test how well HZ deals with nested invocations.
- */
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
 public class InvocationNestedRemoteTest extends InvocationNestedTest {
 
-    private final String response = "someresponse";
+    private static final String RESPONSE = "someresponse";
 
-    // ======================= remote invocations ============================================
+    @Rule
+    public ExpectedException expected = ExpectedException.none();
 
     @Test
-    public void whenPartition_callsGeneric() {
+    public void invokeOnPartition_outerGeneric_innerGeneric_forbidden() {
+        HazelcastInstance[] cluster = createHazelcastInstanceFactory(1).newInstances();
+        HazelcastInstance local = cluster[0];
+        OperationService operationService = getOperationService(local);
+
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, GENERIC_OPERATION);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, GENERIC_OPERATION);
+
+        expected.expect(Exception.class);
+        operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
+    }
+
+    @Test
+    public void invokeOnPartition_outerRemote_innerGeneric() {
         HazelcastInstance[] cluster = createHazelcastInstanceFactory(2).newInstances();
         HazelcastInstance local = cluster[0];
         HazelcastInstance remote = cluster[1];
         OperationService operationService = getOperationService(local);
 
-        InnerOperation innerOperation = new InnerOperation(response, -1);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, GENERIC_OPERATION);
         OuterOperation outerOperation = new OuterOperation(innerOperation, getPartitionId(remote));
-
         InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        assertEquals(response, f.getSafely());
+        assertEquals(RESPONSE, f.getSafely());
     }
 
     @Test
-    public void whenPartition_callsSamePartition() {
+    public void invokeOnPartition_outerRemote_innerSameInstance_samePartition() {
         HazelcastInstance[] cluster = createHazelcastInstanceFactory(2).newInstances();
         HazelcastInstance local = cluster[0];
         HazelcastInstance remote = cluster[1];
         OperationService operationService = getOperationService(local);
 
         int partitionId = getPartitionId(remote);
-        InnerOperation innerOperation = new InnerOperation(response, partitionId);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, partitionId);
         OuterOperation outerOperation = new OuterOperation(innerOperation, partitionId);
-
         InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        assertEquals(response, f.getSafely());
+        assertEquals(RESPONSE, f.getSafely());
     }
 
     @Test
-    public void whenPartition_callsDifferentPartition_butMappedToSameThread() throws ExecutionException, InterruptedException {
+    public void invokeOnPartition_outerRemote_innerSameInstance_callsDifferentPartition_mappedToSameThread() throws ExecutionException, InterruptedException {
         HazelcastInstance[] cluster = createHazelcastInstanceFactory(2).newInstances();
         HazelcastInstance local = cluster[0];
         HazelcastInstance remote = cluster[1];
         OperationService operationService = getOperationService(local);
 
         int outerPartitionId = getPartitionId(remote);
-        int innerPartitionId = 0;
-        for (; innerPartitionId < remote.getPartitionService().getPartitions().size(); innerPartitionId++) {
-            if (innerPartitionId == outerPartitionId) {
-                continue;
-            }
-
-            if (!getPartitionService(remote).getPartition(innerPartitionId).isLocal()) {
-                continue;
-            }
-
-            if (mappedToSameThread(operationService, outerPartitionId, innerPartitionId)) {
-                break;
-            }
-        }
-        InnerOperation innerOperation = new InnerOperation(response, innerPartitionId);
+        int innerPartitionId = randomPartitionIdMappedToSameThreadAsGivenPartitionIdOnInstance(outerPartitionId, remote, operationService);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, innerPartitionId);
         OuterOperation outerOperation = new OuterOperation(innerOperation, outerPartitionId);
-
         InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        assertEquals(response, f.getSafely());
+        expected.expect(IllegalThreadStateException.class);
+        expected.expectMessage("cannot make remote call");
+        f.getSafely();
     }
 
     @Test
-    public void whenPartition_callsIncorrectPartition() {
+    public void invokeOnPartition_outerRemote_innerDifferentInstance_forbidden() {
         HazelcastInstance[] cluster = createHazelcastInstanceFactory(2).newInstances();
         HazelcastInstance local = cluster[0];
         HazelcastInstance remote = cluster[1];
@@ -96,45 +95,61 @@ public class InvocationNestedRemoteTest extends InvocationNestedTest {
 
         int outerPartitionId = getPartitionId(remote);
         int innerPartitionId = getPartitionId(local);
-        InnerOperation innerOperation = new InnerOperation(response, innerPartitionId);
+        assertNotEquals("partitions should be different", innerPartitionId, outerPartitionId);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, innerPartitionId);
         OuterOperation outerOperation = new OuterOperation(innerOperation, outerPartitionId);
-
         InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        try {
-            f.getSafely();
-            fail();
-        } catch (IllegalThreadStateException e) {
-        }
+        expected.expect(IllegalThreadStateException.class);
+        expected.expectMessage("cannot make remote call");
+        f.getSafely();
     }
 
     @Test
-    public void whenGeneric_callsGeneric() {
+    public void invokeOnPartition_outerLocal_innerDifferentInstance_forbidden() {
         HazelcastInstance[] cluster = createHazelcastInstanceFactory(2).newInstances();
         HazelcastInstance local = cluster[0];
         HazelcastInstance remote = cluster[1];
         OperationService operationService = getOperationService(local);
 
-        InnerOperation innerOperation = new InnerOperation(response, -1);
-        OuterOperation outerOperation = new OuterOperation(innerOperation, -1);
+        int outerPartitionId = getPartitionId(local);
+        int innerPartitionId = getPartitionId(remote);
+        assertNotEquals("partitions should be different", innerPartitionId, outerPartitionId);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, innerPartitionId);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, outerPartitionId);
+        InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        InternalCompletableFuture f = operationService.invokeOnTarget(null, outerOperation, getAddress(remote));
-
-        assertEquals(response, f.getSafely());
+        expected.expect(IllegalThreadStateException.class);
+        expected.expectMessage("cannot make remote call");
+        f.getSafely();
     }
 
     @Test
-    public void whenGeneric_callsPartition() {
+    public void invokeOnTarget_outerGeneric_innerGeneric() {
         HazelcastInstance[] cluster = createHazelcastInstanceFactory(2).newInstances();
         HazelcastInstance local = cluster[0];
         HazelcastInstance remote = cluster[1];
         OperationService operationService = getOperationService(local);
 
-        InnerOperation innerOperation = new InnerOperation(response, 0);
-        OuterOperation outerOperation = new OuterOperation(innerOperation, -1);
-
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, GENERIC_OPERATION);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, GENERIC_OPERATION);
         InternalCompletableFuture f = operationService.invokeOnTarget(null, outerOperation, getAddress(remote));
 
-        assertEquals(response, f.getSafely());
+        assertEquals(RESPONSE, f.getSafely());
     }
+
+    @Test
+    public void invokeOnTarget_outerGeneric_innerSameInstance() {
+        HazelcastInstance[] cluster = createHazelcastInstanceFactory(2).newInstances();
+        HazelcastInstance local = cluster[0];
+        HazelcastInstance remote = cluster[1];
+        OperationService operationService = getOperationService(local);
+
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, 0);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, GENERIC_OPERATION);
+        InternalCompletableFuture f = operationService.invokeOnTarget(null, outerOperation, getAddress(remote));
+
+        assertEquals(RESPONSE, f.getSafely());
+    }
+
 }

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/InvocationNestedTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/InvocationNestedTest.java
@@ -1,5 +1,6 @@
 package com.hazelcast.spi.impl.operationservice.impl;
 
+import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.spi.AbstractOperation;
@@ -13,11 +14,13 @@ import java.io.IOException;
 
 public abstract class InvocationNestedTest extends HazelcastTestSupport {
 
-    public static boolean mappedToSameThread(OperationService operationService, int partitionId1, int partitionId2){
+    protected static final int GENERIC_OPERATION = -1;
+
+    public static boolean mappedToSameThread(OperationService operationService, int partitionId1, int partitionId2) {
         OperationServiceImpl operationServiceImpl = (OperationServiceImpl) operationService;
         ClassicOperationExecutor executor = (ClassicOperationExecutor) operationServiceImpl.getOperationExecutor();
-        int thread1 =  executor.toPartitionThreadIndex(partitionId1);
-        int thread2 =  executor.toPartitionThreadIndex(partitionId2);
+        int thread1 = executor.toPartitionThreadIndex(partitionId1);
+        int thread2 = executor.toPartitionThreadIndex(partitionId2);
         return thread1 == thread2;
     }
 
@@ -35,7 +38,6 @@ public abstract class InvocationNestedTest extends HazelcastTestSupport {
 
         @Override
         public void run() throws Exception {
-            System.out.println(Thread.currentThread());
             int partitionId = innerOperation.getPartitionId();
             OperationService operationService = getNodeEngine().getOperationService();
             InternalCompletableFuture f;
@@ -80,7 +82,6 @@ public abstract class InvocationNestedTest extends HazelcastTestSupport {
 
         @Override
         public void run() throws Exception {
-            System.out.println(Thread.currentThread());
         }
 
         @Override
@@ -100,4 +101,38 @@ public abstract class InvocationNestedTest extends HazelcastTestSupport {
             value = in.readObject();
         }
     }
+
+    protected static int randomPartitionIdNotMappedToSameThreadAsGivenPartitionIdOnInstance(HazelcastInstance hz, int givenPartitionId) {
+        int resultPartitionId;
+        for (resultPartitionId = 0; resultPartitionId < hz.getPartitionService().getPartitions().size(); resultPartitionId++) {
+            if (resultPartitionId == givenPartitionId) {
+                continue;
+            }
+            if (!getPartitionService(hz).getPartition(resultPartitionId).isLocal()) {
+                continue;
+            }
+            if (!mappedToSameThread(getOperationService(hz), givenPartitionId, resultPartitionId)) {
+                break;
+            }
+        }
+        return resultPartitionId;
+    }
+
+    protected static int randomPartitionIdMappedToSameThreadAsGivenPartitionIdOnInstance(int givenPartitionId, HazelcastInstance instance, OperationService operationService) {
+        int resultPartitionId = 0;
+        for (; resultPartitionId < instance.getPartitionService().getPartitions().size(); resultPartitionId++) {
+            if (resultPartitionId == givenPartitionId) {
+                continue;
+            }
+            if (!getPartitionService(instance).getPartition(resultPartitionId).isLocal()) {
+                continue;
+            }
+            if (mappedToSameThread(operationService, givenPartitionId, resultPartitionId)) {
+                break;
+            }
+        }
+        return resultPartitionId;
+    }
+
+
 }

